Adds `customElements.polyfillDefineLazy`

sorvell opened the pull request:
This is an optional optimization which allows defining an element with a function that returns a class. This function is called the first time the polyfill tries to upgrade an element of the given name. It allows the polyfill to defer any work needed to generate the element class or its callbacks until this work is actually needed.

Commit:
Adds `polyfillDefineLazy`

This is an optional optimization which allows defining an element with a function that returns a class. This function is called the first time the polyfill tries to upgrade an element of the given name. It allows the polyfill to defer any work needed to generate the element class or its callbacks until this work is actually needed.

Commit:
Merge branch 'prefer-performance' into polyfillDefineLazy2

Commit:
Merge branch 'prefer-performance' into polyfillDefineLazy2

Commit:
Merge branch 'prefer-performance' into polyfillDefineLazy2

bicknellr commented:
Could you merge master again? That should cut down some of the diff.

Commit:
Merge branch 'master' into polyfillDefineLazy2

bicknellr commented on src/CustomElementInternals.js:
```diff
@@ -35,20 +35,82 @@ export default class CustomElementInternals {
     this._hasPatches = false;
 
+    /** @type {boolean} */
+    this.addingDefinitionCallbacks = false;
```
Assuming it's serving the same purpose, can we change this back to `elementDefinitionIsRunning`? This name is from the spec: https://html.spec.whatwg.org/multipage/custom-elements.html#element-definition-is-running

sorvell replied:
Done

bicknellr commented on src/CustomElementInternals.js:
```diff
@@ -35,20 +35,82 @@ export default class CustomElementInternals {
   }
 
+  /**
+   * @param {Object} definition
```
`@param {!CustomElementDefinition} definition`

sorvell replied:
Done

bicknellr commented on src/CustomElementInternals.js:
```diff
@@ -35,20 +35,82 @@ export default class CustomElementInternals {
+    }
+  }
+
+  ensureDefinitionComplete(definition) {
```
All the places where this is called seem to check immediately afterwards if the definition exists, as does this function. What if we change this to `localNameToCompleteDefinition` that would either return undefined if no definition exists or take the existing definition, complete it if necessary, and return the complete one?

sorvell replied:
Done

bicknellr commented on src/CustomElementInternals.js:
```diff
@@ -35,20 +35,82 @@ export default class CustomElementInternals {
+  ensureDefinitionComplete(definition) {
+    if (definition && !definition.isComplete) {
+      try {
+        definition.constructorFunction = definition.constructorFunction();
```
I think that we should move `constructorFunction` to `CustomElementCallbacks` so that it isn't overloaded and replace it with a `constructorGetter` property specifically for that purpose. We should also consider renaming `CustomElementCallbacks` and `CustomElementDefinition`, since `CustomElementCallbacks` wouldn't be just callbacks anymore and `CustomElementDefinition` doesn't contain a full definition.

bicknellr commented on src/CustomElementInternals.js:
```diff
@@ -35,20 +35,82 @@ export default class CustomElementInternals {
+        adoptedCallback,
+        attributeChangedCallback,
+        observedAttributes
+      }
```
`;`, I need to add a formatter to this repo. :/

sorvell replied:
Done

bicknellr commented on externs/custom-elements.js:
```diff
@@ -22,13 +22,21 @@ class AlreadyConstructedMarkerType {}
+ *  isComplete: boolean,
+ *  localName: string,
+ *  constructorFunction: !Function,
+ *  callbacks: (CustomElementCallbacks|undefined),
```
`(!CustomElementCallbacks|undefined)`

sorvell replied:
Done

bicknellr reviewed (commented):
I think `CustomElementDefinition` should probably become a legit class rather than just a struct, given that this PR puts it in a position where it can have invalid states that can't be avoided using the type system. For example, `isComplete` should only be true if `callbacks` is not undefined.

Commit:
Address review feedback

Commit:
Refactoring based on review feedback

* Adds `CustomElementDefinitionProducer` class to support `polyfillDefineLazy`
* Adds documentation to the readme.

bicknellr commented on src/CustomElementDefinitionProducer.js:
```diff
@@ -0,0 +1,136 @@
+let elementDefinitionIsRunning = false;
+
+/**
+ * @unrestricted
```
I don't think this should have `@unrestricted`. (AFAICT, the `@unrestricted` that's already on CustomElementRegistry shouldn't be there either.)

bicknellr commented on src/CustomElementDefinitionProducer.js:
```diff
@@ -0,0 +1,136 @@
+     * @private
+     * @type {!CustomElementDefinition|undefined}
+     */
+    this._definition;
```
`= undefined`

bicknellr commented on src/CustomElementInternals.js:
```diff
@@ -35,6 +35,9 @@ export default class CustomElementInternals {
     this._hasPatches = false;
 
+    /** @type {boolean} */
+    this.elementDefinitionIsRunning = false;
```
This isn't used.

sorvell replied:
Removed.

bicknellr commented on src/CustomElementRegistry.js:
```diff
@@ -142,6 +83,14 @@ export default class CustomElementRegistry {
     }
   }
 
+  define(localName, constructor) {
```
This needs types for the parameters.

sorvell replied:
Done.

bicknellr commented on src/CustomElementRegistry.js:
```diff
@@ -142,6 +83,14 @@ export default class CustomElementRegistry {
+    this._define(localName, constructor, false);
+  }
+
+  polyfillDefineLazy(localName, constructorGetter) {
```
This also needs types.

sorvell replied:
Done.

bicknellr commented on src/CustomElementDefinitionProducer.js:
```diff
@@ -0,0 +1,136 @@
+  /** @return {!CustomElementDefinition} */
+  get definition() {
+    this._ensureDefinition();
+    return /** @type {!CustomElementDefinition} */ (this._definition);
```
This cast can be avoided by making `_ensureDefinition` return the existing definition, if any, or generate the new one and return it and then changing this to `return this._ensureDefinition();`.

bicknellr commented on src/CustomElementDefinitionProducer.js:
```diff
@@ -0,0 +1,136 @@
+    let attributeChangedCallback;
+    let observedAttributes;
+    let constructorFunction;
+    let error;
```
`error` is unused.

bicknellr commented on src/CustomElementDefinitionProducer.js:
```diff
@@ -0,0 +1,136 @@
+      adoptedCallback = getCallback('adoptedCallback');
+      attributeChangedCallback = getCallback('attributeChangedCallback');
+      observedAttributes = constructorFunction['observedAttributes'] || [];
+    } catch (e) {
```
`define` is specced to re-throw exceptions that are thrown from inside any user code it calls<sup>1</sup>, in which case no definition is added to the registry's map.

So, you can do something like this:

```javascript
document.body.appendChild(document.createElement('custom-element'));

try {
  customElements.define('custom-element', class extends HTMLElement {
    constructor() {
      super();
      console.log("First definition constructed.");
    }
    static get observedAttributes() {
      throw new Error('Some error...');
    }
    // `observedAttributes` is only read if `attributeChangedCallback` exists. ^2
    attributeChangedCallback() {}
  });
} catch (e) {
  console.log('oops: ', e);
}

customElements.define('custom-element', class extends HTMLElement {
  constructor() {
    super();
    console.log("Second definition constructed.");
  }
});
```

... and the `<custom-element>` added in the first line will be upgraded with the second definition.

I think this means that this function should still bail when this catch happens and we'll need to update `get definition` to return `{!CustomElementDefinition|undefined}` as well as things that depend on it. Also, `this._internals.localNameToDefinitionProducer(localName)` doesn't completely determine if a definition already exists for an element anymore: you could have an existing producer for a lazy definition that will fail but hasn't been run yet. It would need to both check if the producer exists as well as force the producer to get the definition:

```javascript
const existingProducer = this._internals.localNameToDefinitionProducer(localName);
if (existingProducer && producer.definition) {
  throw new Error(`A custom element with name '${localName}' has already been defined.`);
}
```

We should pull the producer out of the map in CustomElementInternals when it fails in this way.

---

<sup>1</sup> Step 14 of [`define`](https://html.spec.whatwg.org/multipage/custom-elements.html#dom-customelementregistry-define) is run while catching any exceptions, then throws any caught exceptions as the last sub-step. The polyfill is incorrect and it suppresses these errors. This could probably be fixed by changing this `return;` to `throw e;`:
https://github.com/webcomponents/custom-elements/blob/92f66f713dc29e0428855ed7eaf4b1e5e02a3e92/src/CustomElementRegistry.js#L118

<sup>2</sup> The polyfill is incorrect and always reads `observedAttributes`:
https://github.com/webcomponents/custom-elements/blob/92f66f713dc29e0428855ed7eaf4b1e5e02a3e92/src/CustomElementRegistry.js#L116

Commit:
Address review feedback.

All but error handling.

Commit:
Merge branch 'master' into polyfillDefineLazy2

bicknellr commented on src/CustomElementDefinitionProducer.js:
```diff
@@ -0,0 +1,134 @@
+ * subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+import CustomElementInternals from './CustomElementInternals.js';
+ import * as Utilities from './Utilities.js';
```
```suggestion
import * as Utilities from './Utilities.js';
```

bicknellr commented on src/CustomElementDefinitionProducer.js:
```diff
@@ -0,0 +1,134 @@
+export default class CustomElementDefinitionProducer {
+
+  /**
+   *
```
extra line

sorvell replied:
Removed.

bicknellr commented on src/CustomElementDefinitionProducer.js:
```diff
@@ -0,0 +1,134 @@
+      attributeChangedCallback,
+      observedAttributes,
+      constructionStack: []
+    }
```
```suggestion
    };
```

bicknellr commented on tests/js/polyfill-define-lazy.js:
```diff
@@ -0,0 +1,141 @@
+suite('polyfillLazyDefine', function() {
+  var work;
+  var assert = chai.assert;
+  var HTMLNS = 'http://www.w3.org/1999/xhtml';
```
`HTMLNS` is unused here.

sorvell replied:
Removed.

bicknellr commented on src/CustomElementDefinitionProducer.js:
```diff
@@ -0,0 +1,134 @@
+     */
+    this._definition = undefined;
+
+    if (!(constructorOrGetter instanceof Function)) {
```
We should copy this check into `_ensureDefinition` to test `constructorFunction`. Also, in this one, it might also be worth throwing errors with different messages depending on `isGetter` since the current error message still reads way it did when `constructorOrGetter` here was only `constructorFunction`.

bicknellr commented on externs/custom-elements.js:
```diff
@@ -20,6 +20,14 @@ CustomElementRegistry.prototype.shadyDomFastWalk;
 
+/**
+ * @typedef {{
+ * localName: string,
```
```suggestion
 *   localName: string,
```

bicknellr commented on externs/custom-elements.js:
```diff
@@ -20,6 +20,14 @@ CustomElementRegistry.prototype.shadyDomFastWalk;
+/**
+ * @typedef {{
+ * localName: string,
+ * definition: !CustomElementDefinition
```
```suggestion
 *   definition: !CustomElementDefinition
```

Commit:
Throw error when `createElement` does not produce an HTMLElement

Fixes #212.

bicknellr committed:
Use `upgradeElement` in `Document#createElement{,NS}`. Fix incorrectly suppressed error in `upgradeElement`.

bicknellr committed:
Test that an element's constructor returns the element being constructed; fix another test.

Commit:
Merge branch 'fix-createElement-error' into polyfillDefineLazy2

Commit:
Address review feedback

Ensure that lazily defined elements that fail to produce a definition release themselves from the registry so that the element name can be used again. This matches the behavior of a failed define call.

Commit:
Fix error message

Commit:
Update externs/custom-elements.js

Co-Authored-By: Russell Bicknell <bicknellr@gmail.com>

Commit:
Update externs/custom-elements.js

Co-Authored-By: Russell Bicknell <bicknellr@gmail.com>

Commit:
Update src/CustomElementDefinitionProducer.js

Co-Authored-By: Russell Bicknell <bicknellr@gmail.com>

Commit:
Update src/CustomElementDefinitionProducer.js

Co-Authored-By: Russell Bicknell <bicknellr@gmail.com>

Commit:
Address review feedback.

dfreedm commented:
Ported to https://github.com/webcomponents/polyfills/pull/151
